daemon, snapmgr: use the snapmgr inside the daemon

mvo5 opened the pull request:
Please do not merge yet, this is to get feedback.

Some open questions:
- the daemon handle license by passing an interacter (progress.Meter) - this is a open question

pedronis commented on overlord/overlord.go:
```diff
@@ -78,6 +78,7 @@ func New() (*Overlord, error) {
 	}
 	o.snapMgr = snapMgr
 	o.stateEng.AddManager(o.snapMgr)
+	snapMgr.Init(s)
```
we need to decide what to do about Init? anyway calling Init is a stateEng task not of the overlord

niemeyer replied:
What's the deal here?  Have we found a case where we need the state earlier than we expected?

If necessary, I think it's fine to drop Init altogether and have Manager(state *State) instead.

pedronis commented on daemon/api.go:
```diff
@@ -634,21 +638,29 @@ func (inst *snapInstruction) Agreed(intro, license string) bool {
-	// also to do: commands update their output dynamically, as it changes.
-	return nil
+	inst.overlord.SnapManager().Ensure()
+	inst.overlord.SnapManager().Wait()
```
these can only be very temporary. We would need Wait etc at the overlord level also we don't have a way to ask synchronously for an Ensure on the overlord though, the most we have is EnsureBefore which is async

niemeyer replied:
Yes, this will be broken very soon if it's not already broken.

It sounds very strange to have Install being a synchronous task given how much it takes.

To avoid sorting that out now, though, something nasty such as the following would do. But let's please talk about what we want here instead.
```
for {
        status := chg.Status()
        if status == DoneStatus || status == ErrorStatus {
                break
        }
}
```

niemeyer replied:
[edit: updated the function]

pedronis commented:
we are missing calling Overlord.Run() (and Stop) in the daemon atm, that relates to general Ensure support

niemeyer commented on daemon/api.go:
```diff
@@ -634,21 +638,29 @@ func (inst *snapInstruction) Agreed(intro, license string) bool {
-	}
+	state := inst.overlord.StateEngine().State()
+	state.Lock()
+	chg := state.NewChange("install-snap", fmt.Sprintf("Installing %s from %s", inst.pkg, inst.Channel))
```
I suggest this as the format string:
```
format := "Install %q snap"
if inst.Channel != "stable" {
    format = "Install %q snap from %q channel"
}
```

In other words:
* No "ing" termination on summaries.. this is a requested change, not necessarily in progress
* %q rather than %s, so we don't have sentence being accidentally built
* The stable channel is the usual one.. no need to call it out

mvo5 replied:
Good point, fixed.

niemeyer commented on daemon/api.go:
```diff
@@ -634,21 +638,29 @@ func (inst *snapInstruction) Agreed(intro, license string) bool {
 var snappyInstall = snappy.Install
 
 func (inst *snapInstruction) install() interface{} {
+	// FIXME: handle license agreement
```
The TODO here should be above the commented out code at the end. Otherwise it looks like code left out by mistake.

niemeyer commented on daemon/api.go:
```diff
@@ -666,8 +678,15 @@ func (inst *snapInstruction) remove() interface{} {
 	}
+	state := inst.overlord.StateEngine().State()
+	state.Lock()
+	chg := state.NewChange("remove-snap", fmt.Sprintf("Removing %s", inst.pkg))
```
Same comment about the format string.

niemeyer commented on daemon/api.go:
```diff
@@ -666,8 +678,15 @@ func (inst *snapInstruction) remove() interface{} {
 
-	return snappy.Remove(inst.pkg, flags, inst)
+	inst.overlord.SnapManager().Ensure()
+	inst.overlord.SnapManager().Wait()
```
Same comment about the waiting logic.

niemeyer commented on overlord/snapstate/snapmgr.go:
```diff
@@ -31,19 +31,21 @@ import (
 	t := change.NewTask("install-snap", fmt.Sprintf(i18n.G("Installing %q"), snap))
 	t.Set("name", snap)
 	t.Set("channel", channel)
+	t.Set("flags", flags)
```
It seems easier to cook this as:
```
type installState struct {
        Name string `json:"name"`
        ...
}
```
instead of having to marshal/unmarshal one by one.

niemeyer replied:
This also allows us to omit uninteresting information for the particular request. Updated the example with the flags field.

niemeyer commented on overlord/snapstate/snapmgr.go:
```diff
@@ -31,19 +31,21 @@ import (
+func Remove(change *state.Change, snap string, flags snappy.RemoveFlags) error {
 	t := change.NewTask("remove-snap", fmt.Sprintf(i18n.G("Removing %q"), snap))
 	t.Set("name", snap)
+	t.Set("flags", flags)
```
Same..

mvo5 commented:
I updated the branch based on the latest conversation and will keep updating when the overlord.Settle() code arrives. I converted install, update, remove in the daemon so far. I can also do purge, rollback, activate, deactivate using the existing pattern. Or we just wait until its a bit clearer what the best approach is. Both is fine for me.

mvo5 commented:
Fwiw, this branch works in local testing, I can install/remove snaps this way.

pedronis commented on daemon/api.go:
```diff
@@ -631,24 +638,58 @@ func (inst *snapInstruction) Agreed(intro, license string) bool {
-	// TODO: return the log
-	// also to do: commands update their output dynamically, as it changes.
-	return nil
+	inst.overlord.SnapManager().Ensure()
```
when we'lll need the contribution of more than one manager this really needs to become a state.EnsureBefore(0) and we should call overlord.Loop() in daemon.Run(?)

pedronis commented on daemon/api.go:
```diff
@@ -656,9 +697,24 @@ func (inst *snapInstruction) update() interface{} {
 
-	_, err := snappy.Update(inst.pkg, flags, inst)
-	return err
+	inst.overlord.SnapManager().Ensure()
```
same

mvo5 commented:
@pedronis Indeed, thanks a bunch.

mvo5 commented:
All the snapInstructions are now using the state engine.

pedronis commented on daemon/daemon.go:
```diff
@@ -196,6 +196,7 @@ func (d *Daemon) addRoutes() {
 // Start the Daemon
 func (d *Daemon) Start() {
 	d.tomb.Go(func() error {
+		d.overlord.Loop()
```
it starts its own goroutine so I would put it the line before the Go call

mvo5 replied:
Good point, thanks!

niemeyer commented on daemon/api.go:
```diff
@@ -631,24 +638,58 @@ func (inst *snapInstruction) Agreed(intro, license string) bool {
+		chg.State().Lock()
+		status := chg.Status()
+		err := chg.Err()
+		chg.State().Unlock()
```
This is a bit boring to compute unnecessarily inside a busy loop. I suggest something like this below:
```
st := chg.State()
for {
        st.Lock()
        s := chg.Status()
        if s == DoneStatus || s == ErrorStatus {
                defer st.Unlock()
                return chg.Err()
        }
        st.Unlock()
        time.Sleep(250 * time.Millisecond)
}
```

I'll cook a proper waiter that is not a busy loop soon.

niemeyer commented on overlord/snapstate/export_test.go:
```diff
@@ -17,7 +17,6 @@
-// Package snapstate implements the manager and state aspects responsible for the installation and removal of snaps.
 package snapstate
 
 type BackendIF backendIF
```
Can we call this as ManagerBackend and managerBackend, respectively?

mvo5 replied:
Thanks, done.

niemeyer commented on overlord/snapstate/snapmgr.go:
```diff
@@ -84,33 +75,100 @@ func Manager(s *state.State) (*SnapManager, error) {
 	runner.AddHandler("remove-snap", m.doRemoveSnap)
+	runner.AddHandler("purge-snap", m.doPurgeSnap)
+	runner.AddHandler("rollback-snap", m.doRollbackSnap)
+	runner.AddHandler("set-active-snap", m.doSetActiveSnap)
```
`"activate-snap"` and `Activate` instead of "setting it active"?

mvo5 replied:
Yes, done. I assume we are ok with `s.NewTask("activate-snap", false)` when we deactivate it? Or should I make a handler for  `activate` and `deactivate`?

niemeyer commented on overlord/snapstate/snapstate.go:
```diff
@@ -0,0 +1,109 @@
+// Note that the state must be locked by the caller.
+func Install(s *state.State, snap, channel string, flags snappy.InstallFlags) (*state.TaskSet, error) {
+	t := s.NewTask("install-snap", fmt.Sprintf(i18n.G("Installing %q"), snap))
+	t.Set("state", installState{
```
Everything is the state.. I suggest calling those out as "install-state", "update-state", etc.. so that you have the freedom of introducing further data in the future without it feeling awkward.

pedronis replied:
to be honest  I find calling these *State a bit strange, unless it is what we are going to actually store in the state more permanently,   they are more like installParams (also wonder how we will pass them around when we have more than one task if we have to mostly ignore the change itself, just duplicating some of them in tasks?)

mvo5 replied:
Makes sense, thanks. Added.

niemeyer commented on overlord/snapstate/snapstate.go:
```diff
@@ -0,0 +1,109 @@
+
+// SetActive returns a set of tasks for setting a snap active.
+// Note that the state must be locked by the caller.
+func SetActive(s *state.State, snap string, active bool) (*state.TaskSet, error) {
```
s/set activate/activate/ everywhere in the function, assuming you're happy about the change.

niemeyer commented:
LGTM!
